File: php/list_signup_optout.php
<?php
/*
 * An example class for making optin requests to the pure platform
 * PHP Version 5.6+
 * @category Data Gathering
 * @author Matt Barber
 * @created_at 15/09/2015
 */

  class Pure360Http {
      const LIST_URL = 'http://response.pure360.com/interface/list.php';
    /**
     *  Performs a signup to a list
     *  @param accName      string     A string representing the profile name
     *  @param listName     string     A string representing the list name
     *  @param recipient    string     A string representing the contact detail
     *  @param customData   array      An associative array of headers and
     *                                 values for the list
     *  @param doubleOptin  string     A string (boolean) with passively opt
     *                                 recipient in if set to TRUE (i.e. send a
     *                                 double optin email)
     **/
    public function signup (
        $accName,
        $listName,
        $recipient,
        $customData=[],
        $doubleOptin='FALSE'
    )
    {
        $params = [
                'accName' => $accName,
                'listName' => $listName,
                'doubleOptin' => $doubleOptin,
                'successUrl' => 'NO-REDIRECT',
                'errorUrl' => 'NO-REDIRECT'
            ];
        $params = array_merge($params, $customData);
        if ($this->is_email($recipient))
        {
            $params['email'] = $recipient;
        }
        else
        {
            $parms['mobile'] = $recipient;
        }
        return $this->web_request($params, self::LIST_URL);
    }
    /**
     * Opts a recipient out of a whole profile
     * @param accName   string  a string representing the profile name
     * @param recipient    string     A string representing the contact detail
    **/
    public function optout ($accName, $recipient)
    {
        $params = [
                'accName' => $accName,
                'mode' => 'OPTOUT'
            ];
        if ($this->is_email($recipient))
        {
            $params['email'] = $recipient;
        }
        else
        {
            $parms['mobile'] = $recipient;
        }
        return $this->web_request($params, self::LIST_URL);
    }

    /**
     * Checks if the recipient is an email or mobile
     * @param recipient    string     A string representing the contact detail
    **/
    private function is_email ($recipient)
    {
        if(is_numeric($recipient))
        {
            return FALSE;
        }
        else if(preg_match('/[^@]+@[^@]+\.[^@]+/', $recipient) === FALSE)
        {
            return TRUE;
        }
        else
        {
            throw new Exception('Not an email address or mobile number');
        }
    }
    /**
     * This sends the web request using curl to the endpoint
     * with the query parameters
     * @param payload       array   Query parameters
     * @param url           string  End point
    **/
    private function webRequest($payload, $url)
    {
        $curl = curl_init();
        curl_setopt_array($curl, [
            CURLOPT_URL => $url,
            CURLOPT_HEADER => FALSE,
            CURLOPT_RETURNTRANSFER => TRUE,
            CURLOPT_POST => TRUE,
            CURLOPT_POSTFIELDS => $payload
        ]);
        $response = CURL_EXEC($curl);
        if ($response === FALSE)
            throw new exception(CURL_ERROR($curl), CURL_ERRNO($curl));
        curl_close($curl);
        return $response;

    }
  }
?>

File: python/list_signup_optout.py
'''
An example class for signing a recipient up to a list, and opting out of a profile
Python 3.4
category Data Gathering
author Matt Barber
created_at 14/09/2015
'''
# Relies on the requests module - install with pip install requests
# Github : http://docs.python-requests.org/en/latest/
import requests
import re

class Pure360Http():
    LIST_URL = 'http://response.pure360.com/interface/list.php'

    def signup(self, accName, listName, recipient, customData={}, doubleOptin='FALSE'):
        '''Performs a signup to a list
        :param accName, a string representing the profile name
        :param listName, a string representing the list name
        :param recipient, a string representing the recipient contact detail
        :param customData, a dictionary representing the custom fields and values for the list
        :param doubleOptin, a string (bool) will passively opt recipient in if set to TRUE

        :rtype string
        '''
        params = {
            'accName': accName,
            'listName': listName,
            'doubleOptin': doubleOptin,
            'successUrl': 'NO-REDIRECT',
            'errorUrl': 'NO-REDIRECT'
        }
        params.update(customData)
        params.update(
            {'email': recipient} if self.is_email(recipient)
            else {'mobile': recipient}
        )
        return self.__web_request(params, self.LIST_URL)

    def optout(self, accName, recipient):
        '''Opts a recipient out of a whole profile
        :param accName, a string representing the profile name
        :param recipient, a string representing the recipient contact detail

        :rtype string
        '''
        params = {
            'accName': accName,
            'mode': 'OPTOUT'
        }
        params.update(
            {'email': recipient} if self.is_email(recipient)
            else {'mobile': recipient}
        )
        return self.__web_request(params, self.LIST_URL)

    def is_email(self, recipient):
        '''Checks if the recipient is a mobile or email
        :param recipient, a string representing the recipient contact detail

        :rtype boolean
        '''
        try:
            int(recipient)
            return False
        except ValueError:
            if not re.match(r"[^@]+@[^@]+\.[^@]+", recipient):
                raise ValueError('Not an email address or mobile number')
            else:
                return True


    def __web_request(self, payload, url, upload_file=None):
        '''This method sends the web request to the endpoint with the query parameters

        :param payload, a dictionary of key value pairs that make up the query parameters
        :param url, a string representing an endpoint

        :rtype string
        '''
        req = requests.post(url, data=payload)
        return req.text
